src/org/labkey/test/components/ui/entities/EntityInsertPanel.java

```diff
@@ -102,73 +108,105 @@ public EntityInsertPanel addRecords(List<Map<String, Object>> records)
 
     public EntityInsertPanel setRecordValues(Map<String, Object> columnValues)
     {
-        int insertRowIndex = elementCache().grid.getEditableRowIndices().get(0);
+        int insertRowIndex = grid().getEditableRowIndices().get(0);
         return setRecordValues(columnValues, insertRowIndex);
     }
 
     public EntityInsertPanel setRecordValues(Map<String, Object> columnValues, int row)
     {
         for(String columnName : columnValues.keySet())
         {
-            elementCache().grid.setCellValue(row, columnName, columnValues.get(columnName));
+            grid().setCellValue(row, columnName, columnValues.get(columnName));
         }
 
         return this;
     }
 
     public EditableGrid getEditableGrid()
     {
-        return elementCache().grid;
+        showGrid();
+        return grid();
+    }
+
+    public FileUploadPanel getFileUploadPanel()
+    {
+        showFileUpload();
+        return fileUploadPanel();
+    }
+
+    private EditableGrid grid()
+    {
+        return new EditableGrid.EditableGridFinder(_driver).timeout(WAIT_FOR_JAVASCRIPT).waitFor(this);
+    }
```

Comment:
I'm still not clear how this is more advantageous than putting this in the elementCache. It would be the same as the other functions in that class that return an element.

Reply:
It beats infinite recursion.
Or, we could write a waitForLoaded method that doesn't use the grid or the fileUploadPanel and still call it from the elementCache constructor, or we could put the waitForLoaded in the component constructor but that puts the wait up front instead of deferring it until you try to do something with the instance.
Also, first-order getters like these (which don't cache) also don't risk staleness in a component if a test toggles to the other mode the component supports